crates/sui-graphql-rpc/src/extensions/timeout.rs

```diff
@@ -2,35 +2,69 @@
 // SPDX-License-Identifier: Apache-2.0
 
 use async_graphql::{
-    extensions::{Extension, ExtensionContext, ExtensionFactory, NextRequest},
-    Response, ServerError,
+    extensions::{Extension, ExtensionContext, ExtensionFactory, NextExecute, NextParseQuery},
+    parser::types::ExecutableDocument,
+    Response, ServerError, ServerResult,
 };
-use std::sync::Arc;
+use async_graphql_value::Variables;
 use std::time::Duration;
-use tokio::time::timeout;
+use std::{net::SocketAddr, sync::Arc};
+use tokio::{sync::Mutex, time::timeout};
+use tracing::error;
+use uuid::Uuid;
 
-use crate::config::ServiceConfig;
+use crate::{config::ServiceConfig, error::code};
 
-#[derive(Clone, Debug, Default)]
-pub(crate) struct Timeout;
+#[derive(Debug, Default)]
+pub(crate) struct Timeout {
+    pub query: Mutex<Option<String>>,
+}
 
 impl ExtensionFactory for Timeout {
     fn create(&self) -> Arc<dyn Extension> {
-        Arc::new(Timeout)
+        Arc::new(Timeout {
+            query: Mutex::new(None),
+        })
     }
 }
 
 #[async_trait::async_trait]
 impl Extension for Timeout {
-    async fn request(&self, ctx: &ExtensionContext<'_>, next: NextRequest<'_>) -> Response {
+    async fn parse_query(
+        &self,
+        ctx: &ExtensionContext<'_>,
+        query: &str,
+        variables: &Variables,
+        next: NextParseQuery<'_>,
+    ) -> ServerResult<ExecutableDocument> {
+        let document = next.run(ctx, query, variables).await?;
+        *self.query.lock().await = Some(ctx.stringify_execute_doc(&document, variables));
```

Comment:
That is a good question. From my tests, that shouldn't happen, but honestly, I do not know how the precedence of executing these extensions functions. The `query_limits_checker` checks on the query payload size in the same `parse_query`, but I think it only uses the query string and variables, and this extra `query` we use for timeouts is not added to that pile.

Reply:
I think the order in which we add them to the schema is the order in which they are executed. Idk how clean it would be, but we could have the `query_limits_checker` add the stringified doc for the timeout extension to pick up

Reply:
Not sure I follow. This is my mental model:
Extension 1: Timeout
Extension 2: QueryLimitsChecker
Extension 3: Logger

Ext 1 + Ext 2  + Ext 3 ==> all call `parse_query`
Ext 1 + Ext 3 ==> call `execute`. 

I do not understand/know in which order is `parse_query` called. I assume it is called 3 times, once for each extension.

Reply:
Not sure how to find this, but my intuition says that all 3 need to finish, so they the next call (resolve) will have to wait until these are done. If one of them errors, then the whole thing errors and it returns.

And finally, the query payload will not be impacted by this, so it should all be good.

Reply:
Yes, all three parse queries need to finish before `async_graphql` moves on to the next phase, but the question is whether we can guarantee that the the limits check happens before the the timeout check, always, so that we don't try and copy a query that we will shortly find out is too big).